File: gtk/src/models/meta.ts
import { Clock, SystemClock } from "./clock.js";
import { Timestamp } from "./timestamp.js";

export interface MetaFile {
  filename: string;
  filenameWithNanos: string;
  contents: string;
}

export class Meta {
  angaFilename = "";
  note = "";
  clock: Clock = new SystemClock();

  constructor(angaFilename: string, note: string, clock: Clock) {
    this.angaFilename = angaFilename;
    this.note = note;
    this.clock = clock;
  }

  toMetaFile(): MetaFile {
    const timestamp = new Timestamp(this.clock.now());
    // Replace triple single quotes with triple double quotes to prevent TOML parsing issues
    const sanitizedNote = this.note.replace(/'''/g, '"""');
    const tomlContent = `[anga]
filename = "${this.angaFilename}"

[meta]
note = '''${sanitizedNote}'''
`;
    return {
      filename: `${timestamp.plain}-note.toml`,
      filenameWithNanos: `${timestamp.withNanos}-note.toml`,
      contents: tomlContent,
    };
  }
}

File: wpf/src/Kaya.Core/Models/Meta.cs
namespace Kaya.Core.Models;

public record MetaFile(string Filename, string FilenameWithNanos, string Contents);

public class Meta
{
    private readonly string _angaFilename;
    private readonly string _note;
    private readonly IClock _clock;

    public Meta(string angaFilename, string note, IClock clock)
    {
        _angaFilename = angaFilename;
        _note = note;
        _clock = clock;
    }

    public MetaFile ToMetaFile()
    {
        var timestamp = new KayaTimestamp(_clock.Now());
        var sanitizedNote = _note.Replace("'''", "\"\"\"");
        var tomlContent = $"""
            [anga]
            filename = "{_angaFilename}"

            [meta]
            note = '''{sanitizedNote}'''
            """.Replace("            ", "") + "\n";

        return new MetaFile(
            $"{timestamp.Plain}-note.toml",
            $"{timestamp.WithNanos}-note.toml",
            tomlContent
        );
    }
}
